cou/steps/analyze.py

```diff
@@ -0,0 +1,238 @@
+            self.name: {
+                "model_name": self.model_name,
+                "charm": self.charm,
+                "charm_origin": self.charm_origin,
+                "os_origin": self.os_origin,
+                "channel": self.channel,
+                "units": {
+                    unit: {
+                        "workload_version": details.get("workload_version", ""),
+                        "os_version": details.get("os_version", ""),
+                    }
+                    for unit, details in self.units.items()
+                },
+            }
+        }
+        yaml = YAML()
+        with StringIO() as stream:
+            yaml.dump(summary, stream)
+            return stream.getvalue()
+
+    def _get_representative_workload_pkg(self) -> str:
+        """Get the representative package name of a charm workload.
+
+        :return: Package name that represents the charm workload. E.g: cinder-common
+        :rtype: str
+        """
+        try:
+            package = CHARM_TYPES[self.charm]["representative_workload_pkg"]
+        except KeyError:
+            logging.warning(
+                "Representative workload package not found for application: %s", self.name
+            )
+            package = ""
+        return package
+
+    def _get_current_os_version(self, workload_version: str) -> str:
+        """Get the openstack version of a unit.
+
+        :param workload_version: Version of the workload of a charm. E.g: 10.2.6
+        :type workload_version: str
+        :return: OpenStack version detected. If not detected return an empty string.
+            E.g: ussuri.
+        :rtype: str
+        """
+        version = ""
+        package = self._get_representative_workload_pkg()
+
+        if package and workload_version:
+            version = get_os_code_info(package, workload_version)
+        return version
+
+    def _get_os_origin(self) -> str:
+        """Get application configuration for openstack-origin or source.
+
+        :return: Configuration parameter of the charm to set OpenStack origin.
+            E.g: cloud:focal-wallaby
+        :rtype: str
+        """
+        for origin in ("openstack-origin", "source"):
+            if self.config.get(origin):
+                return self.config[origin].get("value", "")
+
+        logging.warning("Failed to get origin for %s, no origin config found", self.name)
+        return ""
+
+    def _get_charm_workload_version(self, unit: str) -> str:
+        """Get the payload version of a charm.
+
+        The workload version of a charm is normally set by a representative debian
+        package. E.g: charm keystone uses keystone package version to represent it.
+
+        :param unit: Unit to detect workload version. E.g: keystone/0
+        :type unit: str
+        :return: Workload version. E.g: 10.1.6
+        :rtype: str
+        """
+        try:
+            return self.status.units[unit].workload_version
+        except AttributeError:
```

Comment:
[iiuc](https://github.com/juju/python-libjuju/blob/d9df8adc84bf97b0388124778dcb27bbf9e9de71/juju/client/_definitions.py#L2929) workload_version defaults to None, so this exception would not be raised.

Reply:
removed function completely